### PR TITLE
feat(a32nx): takeoff performance calculator

### LINKED ISSUE #3393: [REQUEST] V1/VR/V2 speed calculation
**V1/VR/V2 speed calculation error**

The calculation of the V1/VR/V2 speed on the PERF page on the MCDU does not include the calculation of fuel data, so the calculation of the V1/VR/V2 speed is unreasonable.The calculation of the V1/VR/V2 speed on the PERF page on the MCDU does not include the calculation of fuel data, so the calculation of the V1/VR/V2 speed is unreasonable. The data used for calculation should include a variety of necessary or possible factors, such as wind direction, wind speed, air pressure, aircraft load, fuel quantity and so on.

**References**

**Additional context**

Discord username (if different from GitHub):

**BY REMOVING THIS LINE, I ATTEST THAT I HAVE FOLLOWED THE ISSUE TEMPLATE AND PROVIDED A TITLE, as well as searched for other possible duplicate issues. I acknowledge if this line is not removed and/or the issue template is not respected, my issue will be closed without warning**

### LINKED ISSUE #4444: Takeoff calculations on EFB.
**Is your feature request related to a problem? Please describe.**

Hi, it would be great to have a take off performance calculator with flex to temp calculation too and v speeds.
**Describe the feature you'd like to see implemented**

**References**

**Additional context**

Discord username (if different from GitHub):

**BY REMOVING THIS LINE, I ATTEST THAT I HAVE FOLLOWED THE ISSUE TEMPLATE AND PROVIDED A TITLE, as well as searched for other possible duplicate issues. I acknowledge if this line is not removed and/or the issue template is not respected, my issue will be closed without warning**

### PR DESCRIPTION
**Pull request by tracernz**
Fixes #3393
Fixes #4444

## Summary of Changes

- Moves the performance calculators out of fbw-common into the aircraft. The calculator interface is designed to be generic, and the EFB UI pages should handle any aircraft types we have. Aircraft provide their calculators to the EFB through a Context. The pages will not appear in the EFB UI if no calculator is provided by the aircraft.
- Implements a complete takeoff performance calculator for dry, wet, and contaminated runways, roughly equivalent to the FlySmart+ Takeoff app, in the EFB. The calculator does not consider obstacles as we have no obstacle database available.

Notes:
- The takeoff calculator is structured closely following the takeoff calculation documentation so it remains maintainable as that document evolves. It might not be the optimal code structure for other reasons, but this one overrides them all.

The takeoff performance model is developed and maintained by @donstim. Huge thanks. Also a big thanks to everyone who helped with development of that model.

TODO: Check out if anything needs done for localisations.

## Additional context

The takeoff calculator is in shared rather than EFB specifically because I intend to implement takeoff performance uplink in future  (https://github.com/flybywiresim/aircraft/issues/8667) and it may also be useful/extended for TOS2. Note that TO perf uplink in the FMS **does not** uplink a single perf result (e.g. from the EFB), but rather a series of results for the selected runways and the intent is to implement like that as in the real thing.

Discord username (if different from GitHub):

## Testing instructions

Test the general usability and buginess, and any questionable results. The takeoff model has been validated on a wide range of scenarios so it's mostly about the UI and interactions, and rejecting invalid cases.

## How to download the PR for QA

Every new commit to this PR will cause new A32NX and A380X artifacts to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, find and click on the **PR Build** tab
1. Click on either **flybywire-aircraft-a320-neo** or **flybywire-aircraft-a380-842** download link at the bottom of the page

### CODE REVIEW

File: fbw-a32nx/src/systems/instruments/src/EFB/index.tsx
Comment (2hwk) [inline fbw-a32nx/src/systems/instruments/src/EFB/index.tsx:9]: This namespacing is really ugly but too bad, it will have to be like this for now.

File: fbw-a32nx/src/systems/shared/src/performance/a32nx_takeoff.ts
Comment (2hwk) [inline fbw-a32nx/src/systems/shared/src/performance/a32nx_takeoff.ts:3354]: May want to elaborate on what properly means in this context :P

File: fbw-common/src/systems/instruments/src/EFB/Localization/data/en.json
Comment (2hwk) [inline fbw-common/src/systems/instruments/src/EFB/Localization/data/en.json:408]: note for maintainers: check that localazy strings have been added before merging

**Review by 2hwk — CHANGES_REQUESTED** [review]

File: fbw-a32nx/src/systems/shared/src/performance/a32nx_takeoff.ts
Comment (tracernz) [reply fbw-a32nx/src/systems/shared/src/performance/a32nx_takeoff.ts:3354]: A proper deep copy that doesn't just serialise and de-serialise. 💀

File: fbw-a32nx/src/systems/instruments/src/EFB/index.tsx
Comment (tracernz) [reply fbw-a32nx/src/systems/instruments/src/EFB/index.tsx:9]: Open to suggestions. It is a landing calculator for that exact type.

Comment (2hwk) [reply fbw-a32nx/src/systems/instruments/src/EFB/index.tsx:9]: Root of the problem is, ideally we'd want the EFB be generic for all FBW a/c types, so we would just have a generic landing calculator FC LandingCalculator that doesn't need awareness of what the variant the plane is, and some input params to that that would be defined per variant (maybe using the VFS config OR xmlConfig (Note: I have no idea how to import XML config into react (probably not possible as we do not inherit BaseInstruments in react displays). 

Most likely we'd have to wait until after flyPad OS 4 (4.1? 5?) to properly abstract this away.

Comment (tracernz) [reply fbw-a32nx/src/systems/instruments/src/EFB/index.tsx:9]: That generic interface is already implemented here. This is the aircraft-specific context that provides the calculator (through the generic interface) to the generic EFB.

Comment (tracernz) [reply fbw-a32nx/src/systems/instruments/src/EFB/index.tsx:9]: This is literally the point of commit 2ccf15c3a9b387b91b4183e8b23bcfb6f8d0ac6e.

File: fbw-common/src/systems/instruments/src/EFB/Localization/data/en.json
Comment (tracernz) [reply fbw-common/src/systems/instruments/src/EFB/Localization/data/en.json:408]: I've posted the keys in the #localisation channel now. 👍

**Review by Saschl — APPROVED** [review]

**Comment by xmudrii** [comment]
Non-QA Team Report (asked by @tracernz)

Discord: xmudrii
Object of testing: #8678
Tier of Testing: 2
Date: 27/6/2024

Testing Process:

- Spawned at a ramp, turned on EFB, opened the calculator page
- Tried to calculate takeoff performance for different:
  - airports (e.g. LYBE, LGTS, LGAV, LYPG, EDDM, EDDF...)
  - runaways and intersections (different TORA, too high and too low values such as 4000m and 1200m)
  - runway conditions and weather
  - aircraft configuration (FLEX or TOGA, packs on/off, all anti-ice options)
- Tried to calculate takeoff performance for impossible takeoff profiles
- Tried different abnormal values (too high/low temperature, wind, runway parameters)
- Did several full flights from different airports and configurations using calculations provided by the EFB
- Ensured that EFB is fully functional after adding the new feature and that other calculators are working

Negatives:

- There are some issues with the flyPadOS that has been reported on Discord. A follow-up issue will be created later

Testing Results:
Passed

Conclusions:
LGTM

**Comment by alepouna** [comment]
QA Report

Discord: alepouna
Object of testing: #8678
Tier of Testing: 1
Date: 27/6/2024

Testing Process:
- Spawn at a gate, runway and test various data configurations, all good
- Attempt to "break" the system by using abnormal & erroneous values, all good
- Attempt to calculate values for impossible take off conditions, all good

Testing Results:
Passed

**Pull request merged**
